pkg/acs/calacs/acsccd/blev_funcs_postsm4.c

```diff
@@ -36,152 +39,218 @@ static int make_amp_array(const int arr_rows, const int arr_cols, SingleGroup *
 static int unmake_amp_array(const int arr_rows, const int arr_cols, SingleGroup * im,
                             int amp, double * array);
 
+/* Remove the signal dependent bias shift from post-SM4 full frame WFC data
+ * (based on ISR http://www.stsci.edu/hst/acs/documents/isrs/isr1202.pdf),
+ * as well as from supported subarray data (~post-mid-2016) as identified by
+ * specific aperture names 
+ * (based on ISR http://www.stsci.edu/hst/acs/documents/isrs/isr1703.pdf).
+ *
+ * chip2 consists of amps C & D, chip1 consists of amps A & B. 
+ *
+ * Because full frame data uses both chips, but supported subarray data
+ * is only from a single amp (therefore, a portion of a single chip), this
+ * routine has a variable number of input SingleGroups.
+ *
+ * 16-May-2018 M.D. De La Pena: Generalized bias_shift_corr() to handle subarray data.
+ */
+int bias_shift_corr(ACSInfo *acs, int nGroups, ...) {
+
+  /* Code to handle variable number of input parameters.  The number of SingleGroup is 2 when 
+     processing a full frame (2 chips - all 4 amps) and only 1 when processing a supported subarray 
+     (1 chip from a single amp).
+  */
+  SingleGroup *sg[nGroups];
+  va_list arguments;
+
+  /* Initializing arguments to store all the values being passed to this routine via "..." */
+  va_start (arguments, nGroups);
+  {
+       /* sg[0] is chip2 and sg[1] is chip 1 for a full frame image */
+       {unsigned int i;
+       for (i = 0; i < nGroups; i++) {
+           sg[i]= va_arg (arguments, SingleGroup *);
+       }}
+  }
+  va_end (arguments);
 
-/* remove signal dependent bias shift from post-SM4 full frame WFC data.
- * based on ISR http://www.stsci.edu/hst/acs/documents/isrs/isr1202.pdf
- * chip2 is amps C & D, chip1 is amps A & B. */
-int bias_shift_corr(ACSInfo *acs, SingleGroup *chip2, SingleGroup *chip1) {
   extern int status;
-  
-  int i, j, k;     /* iteration variables */
-  
-  /* array for amp data and amp data + gap pixels */
-  double * ampdata, * ampdata_gap;
-  
+
   const double serial_freq = 1000./22.;    /* serial pixel frequency */
   const double parallel_shift = 3.212;     /* parallel shift time */
   
   /* number of virtual pixels at end of each row */
   const int ngap_pix = (int) (serial_freq * parallel_shift + 0.5);
   
-  const int arr_rows = chip2->sci.data.ny;
-  const int arr_cols = chip2->sci.data.nx/2;
+  /* There is always at least one SingleGroup - use sg[0] to get the image size of a single amp.
+   * If this is a full frame image, then "nx" represents all the columns across both amps plus
+   * the physical overscan and must be divided by 2.  If this is a subarray, then "nx" represents
+   * just the columns for a single amp plus physical overscan.
+   */
+  const int arr_rows = sg[0]->sci.data.ny;
```

Comment:
``sg[0]`` is being accessed without first being checked. This can be done here, or better still when first assigned to ``sg`` so as to shrink it to only valid ptrs (bookkeeping required of ``nGroups`` etc).